zebra-consensus/src/block/check.rs

```diff
@@ -53,6 +57,43 @@ pub fn is_equihash_solution_valid(header: &Header) -> Result<(), equihash::Error
 /// accepted." [§7.5][7.5]
 ///
 /// [7.5]: https://zips.z.cash/protocol/protocol.pdf#blockheader
-pub fn is_time_valid_at(header: &Header, now: DateTime<Utc>) -> Result<(), BoxError> {
+pub fn time_is_valid_at(header: &Header, now: DateTime<Utc>) -> Result<(), BoxError> {
     header.is_time_valid_at(now)
 }
+
+/// [3.9]: https://zips.z.cash/protocol/protocol.pdf#subsidyconcepts
+pub fn subsidy_is_correct(network: Network, block: &Block) -> Result<(), BlockError> {
+    let height = block.coinbase_height().ok_or(SubsidyError::NoCoinbase)?;
+    let coinbase = block.transactions.get(0).ok_or(SubsidyError::NoCoinbase)?;
+
+    let halving_div = subsidy::general::halving_divisor(height, network);
+
+    // TODO: the sum of the coinbase transaction outputs must be less than or equal to the block subsidy plus transaction fees
+
+    // Check founders reward and funding streams
+    if height < SLOW_START_INTERVAL {
+        unreachable!(
+            "unsupported block height: callers should handle blocks below {:?}",
+            SLOW_START_INTERVAL
+        )
+    } else if halving_div.count_ones() != 1 {
+        unreachable!("invalid halving divisor: the halving divisor must be a non-zero power of two")
+    } else if halving_div == 1 {
+        // validate founders reward
+        let founders_reward = subsidy::founders_reward::founders_reward(height, network)
+            .expect("invalid Amount: founders reward should be valid");
+        let matching_values = subsidy::general::find_output_with_amount(coinbase, founders_reward);
+
+        // TODO: the exact founders reward value must be sent as a single output to the correct address
+        if !matching_values.is_empty() {
+            Ok(())
+        } else {
+            Err(SubsidyError::FoundersRewardNotFound)?
+        }
+    } else if halving_div == 2 {
+        unimplemented!("funding stream block subsidy validation is not implemented")
```

Comment:
These conditions are wrong, they should be:
* founders reward: before Canopy activation
* funding streams: after Canopy activation, but before the second halving

See https://github.com/zcash/zips/blob/master/zip-0214.rst#specification , particularly the note about testnet funding streams before the first halving.

We should add a comment about how ZIP-214 overrides ZIP-1014 for testnet.

Reply:
Fixed in 6ef71ba0bbe0f1cfd9d60969a1c4ab876dfe4ccd.